src/lib/spinel/radio_spinel_impl.hpp

```diff
@@ -252,14 +252,17 @@ template <typename InterfaceType> void RadioSpinel<InterfaceType>::ResetRcp(bool
     mIsReady    = false;
     mWaitingKey = SPINEL_PROP_LAST_STATUS;
 
-    if (aResetRadio && (SendReset(SPINEL_RESET_STACK) == OT_ERROR_NONE) && (WaitResponse(false) == OT_ERROR_NONE))
+    if (aResetRadio)
     {
+        SuccessOrExit(SendReset(SPINEL_RESET_STACK));
```

Comment:
In existing code, if `SendReset()` or `WaitResponse()` fails, we would continue after the `if` block to do a hardware set, which we don't in the code flow? I think the existing code is the intended behavior.

Reply:
You are correct this is a bit off. But the problem with the original is that `SendReset` is Sending HDLC crame and resetting the interface between Host and RCP. https://github.com/openthread/openthread/blob/main/src/posix/platform/hdlc_interface.cpp#L216
This connection reset is actually clearing buffers in the host OS potentially malforming HDLC frames.

The previous implementation worked when the `WaitResponse` returned no error.

Reply:
@abtink the last sentence mentions this https://github.com/openthread/openthread/pull/9278/files PR.

Reply:
@zhanglongxia Thanks for reviewing.
I do agree with You that the original logic is similar or even the same as old one. 

But during this refactoring #8858   `WaitResponse`
still had a bug with the rutn error always set to `OT_ERROR_NONE` please refer to : https://github.com/openthread/openthread/pull/8858/files#diff-f499bfcd3d5488045b07a03a07615ffa319ae7af6f193cf0e237674c4f407a8eR1741

So the demon wasn't dying when it reached the `WaitResponse` call. And it was reaching it if the frame was sent correctly to RCP and Reset connection (essentially closing the interface, opening it again clearing the system buffers) was ok.  Meaning no indication of reset had to reach the demon to claim that the reset was ok.

Now THe code is waiting for an indication of the last status from the RCP like the one sent here https://github.com/openthread/openthread/blob/main/src/ncp/ncp_base.cpp#L1212 . So this looks ok unless we hit the race condition of clearing the system buffer while the transmission happens.

As @abtink pointed the code that I wrote needs to be changed. But also the original logic.

Reply:
@canisLupus1313  The `WaitResponse ` is waiting for the response from [ncp_base.cpp#L857](https://github.com/openthread/openthread/blob/main/src/ncp/ncp_base.cpp#L857), rather than [ncp_base.cpp#L1212](https://github.com/openthread/openthread/blob/main/src/ncp/ncp_base.cpp#L1212). After the RCP is restarted, it sends out the RESET reason to the host automatically.

I think the root cause of your issue is that: After the RCP is restarted, it sends out the RESET reason to the host. However, the host is  [re-opening](https://github.com/openthread/openthread/blob/main/src/posix/platform/hdlc_interface.cpp#L724) the UART. Then the RESET reason from the RCP is lost. In the [original code](https://github.com/openthread/openthread/pull/8858/files#diff-f499bfcd3d5488045b07a03a07615ffa319ae7af6f193cf0e237674c4f407a8eR1741), the  `WaitResponse` always return `OT_ERROR_NONE `. There is no problem so far. But the [PR9278](https://github.com/openthread/openthread/pull/9278) changes the return error code, which causes your issue. So the key is that the RESET reason is lost.

The [kOpenFileDelay](https://github.com/openthread/openthread/blob/main/src/posix/platform/hdlc_interface.hpp#L238) is set to 500ms by default. It means the host has to wait `500ms` when failed to open the UART. During the waitting time, the RESET reason may be missed. You can try to set the `kOpenFileDelay` to 1 to check if the host will miss the RESET reason.

Reply:
@zhanglongxia Thanks for this `kOpenFileDelay` hint I will try it and get back to You.

Reply:
@zhanglongxia @abtink looks like setting `kOpenFileDelay` to 1 helps in my case. What Do you say that I will put as a parameter with a default value of 500ms and remove this part of the code?

I have one more question though I can see that `mError`  is set to `OT_ERROR_RESPONSE_TIMEOUT` but I don't see a place where it's set to `OT_ERROR_NONE`  in between two `WaitResponse` calls. Am I missing something?

Reply:
I think you may change the default value of `kOpenFileDelay` to an appropriate value and keep this part of code unchanged.

The error code `mError` is set to `OT_ERROR_NONE` in the function `HandleWaitingResponse()`.